src/reflect/scala/reflect/internal/tpe/TypeMaps.scala

```diff
@@ -433,8 +433,12 @@ private[internal] trait TypeMaps {
     (pre eq NoType) || (pre eq NoPrefix) || !isPossiblePrefix(clazz)
     )
 
-  def newAsSeenFromMap(pre: Type, clazz: Symbol): AsSeenFromMap =
-    new AsSeenFromMap(pre, clazz)
+  final def newAsSeenFromMap(pre: Type, clazz: Symbol): AsSeenFromMap = {
+    val prePackageObjectExpanded = if (pre.typeSymbolDirect.hasPackageFlag && !clazz.hasPackageFlag)
+      pre.packageObject.typeOfThis
+    else pre
```

Comment:
It might be more clean to handle this directly in the `class AsSeenFromMap`, no? Also, since the override in interactive is gone now, could we just get rid of the factory method, and also make the `class AsSeenFromMap` final?